pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/OffHeapStringStore.java

```diff
@@ -0,0 +1,242 @@
+      if (size >= Integer.MAX_VALUE) {
+        size = Integer.MAX_VALUE - 1;
+      }
+      _pinotDataBuffer = PinotDataBuffer.allocateDirect(size);
+      _pinotDataBuffer.order(ByteOrder.nativeOrder());
+      _byteBuffer = _pinotDataBuffer.toDirectByteBuffer(0, (int) size);
+      _startIndex = startIndex;
+      _availEndOffset = _byteBuffer.capacity();
+      _size = size;
+    }
+
+    private int add(String string) {
+      int startOffset = _availEndOffset - string.length() * CHAR_SIZE;
+      if (startOffset <= (_numStrings + 1) * INT_SIZE) {
+        // full
+        return -1;
+      }
+      for (int i = 0, j = startOffset; i < string.length(); i++, j = j + CHAR_SIZE) {
+        _byteBuffer.putChar(j, string.charAt(i));
+      }
+      _byteBuffer.putInt(_numStrings * INT_SIZE, startOffset);
+      _availEndOffset = startOffset;
+      return _numStrings++;
+    }
+
+    private boolean equalsStringAt(String string, int index) {
+      int startOffset = _byteBuffer.getInt(index * INT_SIZE);
+      int endOffset = _byteBuffer.capacity();
+      if (index > 0) {
+        endOffset = _byteBuffer.getInt((index - 1) * INT_SIZE);
+      }
+      if ((endOffset - startOffset)/CHAR_SIZE != string.length()) {
+        return false;
+      }
+      for (int i = 0, j = startOffset; i < string.length(); i++, j = j + CHAR_SIZE) {
+        if (string.charAt(i) != _byteBuffer.getChar(j)) {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private String get(final int index) {
+      int startOffset = _byteBuffer.getInt(index * INT_SIZE);
+      int endOffset = _byteBuffer.capacity();
+      if (index > 0) {
+        endOffset = _byteBuffer.getInt((index - 1) * INT_SIZE);
+      }
+      char[] chars = new char[(endOffset - startOffset)/CHAR_SIZE];
+      for (int i = 0, j = startOffset; i < chars.length; i++, j = j + CHAR_SIZE) {
+        chars[i] = _byteBuffer.getChar(j);
+      }
+      return new String(chars);
+    }
+
+    @Override
+    public void close()
+        throws IOException {
+      _pinotDataBuffer.close();
+    }
+
+    private long getSize() {
+      return _size;
+    }
+
+    private int getStartIndex() {
+      return _startIndex;
+    }
+  }
+
+  private volatile List<Buffer> _buffers = new ArrayList<Buffer>();
+  private int _numElements = 0;
+  private volatile Buffer _currentBuffer;
+
+  public OffHeapStringStore() {
+    expand(START_SIZE);
+  }
+
+  // Expand the buffer size, allocating a min of 32k for strings.
+  private Buffer expand(long size) {
+    Buffer buffer = new Buffer(size, _numElements);
+    List<Buffer> newList = new LinkedList<>();
+    for (Buffer b : _buffers) {
+      newList.add(b);
+    }
+    newList.add(buffer);
+    _buffers = newList;
+    _currentBuffer = buffer;
+    return buffer;
+  }
+
+  private Buffer expand() {
+    Buffer lastBuffer = _buffers.get(_buffers.size()-1);
+    Buffer newBuffer = expand(lastBuffer.getSize());
```

Comment:
Expand buffers with a multiplicative factor, as to avoid degradation due to large numbers of buffers. This also prevents catastrophically allocating all of the memory if someone tries to add a String that's larger than your buffer size.